### PR TITLE
riscv: pmp: enable stackguard without multithreading

### PR DESCRIPTION
**Pull request by f0rget-the-sad**
Without multithreading only two stacks present: ISR and main. As any stack they also could overflow, so it make sense to add stack guard for them also.

Remove stack guard dependency on multithreading and mark `Z_RISCV_STACK_GUARD_SIZE` bytes at the beginning of stack as read-only region with PMP entry.

### CODE REVIEW

**Review by cfriedt — DISMISSED** [review]

File: arch/riscv/core/fatal.c [hunk 1]
```diff
@@ -172,11 +172,20 @@ static bool bad_stack_pointer(struct arch_esf *esf)
...
+#if CONFIG_MULTITHREADING
 	if (sp >= _current->stack_info.start - K_KERNEL_STACK_RESERVED &&
 	    sp <  _current->stack_info.start - K_KERNEL_STACK_RESERVED
 		  + Z_RISCV_STACK_GUARD_SIZE) {
 		return true;
 	}
+#else
+	uintptr_t isr_stack = (uintptr_t)z_interrupt_stacks;
+	uintptr_t main_stack = (uintptr_t)z_main_stack;
+	if ((sp >= isr_stack && sp < isr_stack + Z_RISCV_STACK_GUARD_SIZE) ||
```
Comment (ycsin) [inline arch/riscv/core/fatal.c:184]: please fix:
```suggestion

	if ((sp >= isr_stack && sp < isr_stack + Z_RISCV_STACK_GUARD_SIZE) ||
```

**Review by ycsin — DISMISSED** [review]

File: arch/riscv/core/fatal.c (hunk 1 above)
Comment (f0rget-the-sad) [reply arch/riscv/core/fatal.c:185]: Fixed.

**Comment by cfriedt** [comment]
We need to have some guarantee that this code  has test coverage.

**Review by fkokosinski — CHANGES_REQUESTED** [review]
The code itself LGTM, but I agree with @cfriedt - we'd need to have some testing for this.

There're at least 3 places where we could add/modify a test:
1. `tests/kernel/threads/no-multithreading` - it's already enabled on RISC-V QEMU-based targets, but I'm not sure if this is the right place to add memory protection related tests. Probably not.
2. `tests/kernel/mem_protect/mem_protect` - we could add a RISC-V specific no multithreading testcase here
3. we could add a separate arch-specific test in e.g. `tests/arch/riscv/pmp` (ARM has one in `zephyr/tests/arch/arm/arm_mem_protect`).

It should probably be both 2 and 3.

**Comment by f0rget-the-sad** [comment]
> The code itself LGTM, but I agree with @cfriedt - we'd need to have some testing for this.
> 
> There're at least 3 places where we could add/modify a test:
> 
>     1. `tests/kernel/threads/no-multithreading` - it's already enabled on RISC-V QEMU-based targets, but I'm not sure if this is the right place to add memory protection related tests. Probably not.
> 
>     2. `tests/kernel/mem_protect/mem_protect` - we could add a RISC-V specific no multithreading testcase here
> 
>     3. we could add a separate arch-specific test in e.g. `tests/arch/riscv/pmp` (ARM has one in `zephyr/tests/arch/arm/arm_mem_protect`).
> 
> 
> It should probably be both 2 and 3.

Thank you for pointing in right direction, I'll work on extending the test cases in next commit.

File: tests/arch/riscv/pmp/no-multithreading/testcase.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  platform_allow:
+    - qemu_riscv32
+    - qemu_riscv32e
+    - qemu_riscv64
+  filter: CONFIG_RISCV_PMP
+  ignore_faults: true
+
+tests:
+  arch.riscv.pmp.no-mt.isr-stack-guard:
+    extra_args: EXTRA_CPPFLAGS=-DVIA_TWISTER=0
```
Comment (fkokosinski) [inline tests/arch/riscv/pmp/no-multithreading/testcase.yaml:11]: ... but if we want to go this route, then I'd rather use `EXTRA_CFLAGS` here and change `VIA_TWISTER` to something sounding less generic

**Review by fkokosinski — COMMENTED** [review]
Thanks a lot for adding the test

File: tests/arch/riscv/pmp/no-multithreading/src/main.c [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+#ifdef VIA_TWISTER
+#define PMP_TEST_FUNC_IDX VIA_TWISTER
+#else
+#define PMP_TEST_FUNC_IDX 0
+#endif
+	pmp_test_func[PMP_TEST_FUNC_IDX]();
+
+	zassert_unreachable("Write to stack guard did not fault");
+	TC_END_REPORT(TC_FAIL);
+}
```
[the remark this answers is not in the document]
Comment (f0rget-the-sad) [reply tests/arch/riscv/pmp/no-multithreading/src/main.c]: Write to PMP protected region will cause fault, so second testcase in same suite will not run.
Current approach is based on: tests/kernel/fatal/no-multithreading

**Review by peter-mitsis — APPROVED** [review]

File: tests/arch/riscv/pmp/no-multithreading/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  platform_allow:
+    - qemu_riscv32
+    - qemu_riscv32e
+    - qemu_riscv64
+  filter: CONFIG_RISCV_PMP
+  ignore_faults: true
+
+tests:
+  arch.riscv.pmp.no-mt.isr-stack-guard:
+    extra_args: EXTRA_CFLAGS=-DPMP_TEST_FUNC_IDX=0
```
Comment (ycsin) [inline tests/arch/riscv/pmp/no-multithreading/testcase.yaml:11]: could have used Kconfig instead?

Comment (f0rget-the-sad) [reply tests/arch/riscv/pmp/no-multithreading/testcase.yaml:11]: I think Kconfig option usually meant to allow user to tweak something.
Here it's already possible to select different tests by name, so I don't see benefits in using Kconfig.

**Review by cfriedt — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
Hi @f0rget-the-sad, unfortunately this PR broke the project CI run on main (https://github.com/zephyrproject-rtos/zephyr/actions/runs/11934344226), the fix is not obvious to me so I've opened a revert PR https://github.com/zephyrproject-rtos/zephyr/pull/81673.

Would you be able to troubleshoot and re-submit a new version that works in CI? (or open a hotfix if that's obvious to you).

Apologies for the extra work.

**Comment by fabiobaltieri** [comment]
> I guess main CI runs even more tests, so question is: how do I trigger the main CI before merge?

Yeah the test sets can be different, you can get the list of failed tests from https://github.com/zephyrproject-rtos/zephyr/runs/33270031021 and run those manually, for example for `qemu_riscv64/qemu_virt_riscv64:tests/kernel/fatal/no-multithreading/kernel.no-mt.arbitrary_reason` you can use 

```console
$ west build -p -b qemu_riscv64 -T tests/kernel/fatal/no-multithreading/kernel.no-mt.arbitrary_reason
$ west build -t run
...
 FAIL - test_fatal in 0.000 seconds
===================================================================
===================================================================
PROJECT EXECUTION FAILED
```

**Comment by f0rget-the-sad** [comment]
Ok, I manually tested tests/kernel/fatal/no-multithreading, tests/kernel/threads/no-multithreading and tests/lib/multi_heap (seems like all failures reported by main CI)with this hotfix: https://github.com/zephyrproject-rtos/zephyr/pull/81676
Let me know if there are other failures.

**Comment by fabiobaltieri** [comment]
@f0rget-the-sad thanks for the quick turnaround.

**Pull request merged**